File: Chapter 02/BuilderDesignPatternExample.java
class Truck {
    private String engine;
    private String color;
    private int seating;

    private Truck() {}

    public String getEngine() {
        return engine;
    }
    public String getColor() {
        return color;
    }
    public int getSeating() {
        return seating;
    }

    
    static class Builder {
        private Truck truck = new Truck();
        public Builder withEngine(String engine) {
            truck.engine = engine;
            return this;
        }
        public Builder withColor(String color) {
            truck.color = color;
            return this;
        }
        public Builder withSeating(int seating) {
            truck.seating = seating;
            return this;
        }
        public Truck build() {
            return truck;
        }
    }
}


public class BuilderDesignPatternExample {
    public static void main(String[] args) {
        Truck truck = new Truck.Builder()
                .withEngine("EcoBoost")
                .withColor("White Gold")
                .withSeating(6)
                .build();

        System.out.println("Truck Configuration:");
        System.out.println("Engine: " + truck.getEngine());
        System.out.println("Color: " + truck.getColor());
        System.out.println("Seating: " + truck.getSeating());
    }
}

File: Chapter 02/BuilderDesignPatternExample.py
class Truck:
    def __init__(self):
        self.engine = None
        self.color = None
        self.seating = None
    def __str__(self):
        return f"Truck Configuration:\nEngine: {self.engine}\nColor: {self.color}\nSeating: {self.seating}

class TruckBuilder:
    def __init__(self):
        self.truck = Truck()
    def with_engine(self, engine):
        self.truck.engine = engine
        return self
    def with_color(self, color):
        self.truck.color = color
        return self
    def with_seating(self, seating):
        self.truck.seating = seating
        return self
    def build(self):
        return self.truck

if __name__ == "__main__":
    builder = TruckBuilder()
    truck = builder.with_engine("EcoBoost")\
                     .with_color("White Gold")\
                     .with_seating(6)\
                     .build()
    print(truck)
